Marketplace: Adds Categories Responsive Menu

Rolilink opened the pull request:
#### Screencast
![Screen Capture on 2022-05-16 at 20-28-33](https://user-images.githubusercontent.com/1035546/168709024-1b1c9f89-dc94-4431-a67e-4732758a7c55.gif)

#### Changes proposed in this Pull Request

* Adds the `ResponsiveToggleGroup` to `@automattic/components`.
* Adds the responsive categories menu into the marketplace.

#### Testing instructions

* Navigate into the marketplace
* Test multiple categories, check everything is working fine.

Related to #62891

Rolilink committed:
added basic intersection functionality

Rolilink committed:
cleaning up a little before implementing the "more" dropdown

Rolilink committed:
using props for hide and show ratio

Rolilink committed:
updating interceptionCallback dependencies array

Rolilink committed:
Added dropdown menu

Rolilink committed:
rendering grouped items

Rolilink committed:
improved styles

Rolilink committed:
cleaning up and centralizing children getter
rendering selected item on more label

Rolilink committed:
polish: removes selected item from group

Rolilink committed:
fine tuning a little more

Rolilink committed:
adding initialActiveIndex prop
adds and cleanup stories

Rolilink committed:
fixes issues with icon not showing
only shows more as label of more ...

Rolilink committed:
improving types

Rolilink committed:
1.1.0

Rolilink committed:
integrating ResponsiveToolbarGroup into the marketplace

Rolilink committed:
fixes navigation issues

Rolilink committed:
use @wordpress/icons for the icon

Rolilink committed:
missing this code removal

Rolilink committed:
fine tuning styling

Rolilink committed:
hidding categories on paid and top free plugins

Rolilink commented:
This one still requires some tests + docs and some design input

Rolilink committed:
resetting version

cpapazoglou commented on packages/components/src/responsive-toolbar-group/index.tsx:
```diff
@@ -0,0 +1,215 @@
+		shadowListItems[ index ] = element;
+	};
+
+	const getChildrenToRender = () =>
+		Object.keys( groupedIndexes ).map( ( index ) => ( {
+			index,
+			grouped: groupedIndexes[ index ],
+			child: children[ parseInt( index ) ],
+		} ) );
```
Can this or any other function be moved outside the function component? I don't see any dependency and we would avoid declaring them again on every render.

Rolilink replied:
Sounds good will take a look

cpapazoglou commented:
Thanks @Rolilink for working on this. Looks good! Testing it seems that accessibility-wise it works fine!

1. On smaller viewports, I think it would be preferable having a scrolling menu?
![CleanShot 2022-05-17 at 09 51 33@2x](https://user-images.githubusercontent.com/12430020/168747733-558663f1-2048-453f-9c2c-43be99cf5ba6.png)

2. Let's make sure that the navigation header is always on top
<img width="574" alt="CleanShot 2022-05-17 at 10 01 03@2x" src="https://user-images.githubusercontent.com/12430020/168749247-1c8ac2d3-e9cb-4b53-b2d3-7ef4824cb75b.png">

3. I really like the animation on selection 🎉 ! Have you considered adding a hover effect too?

4. For the future:
a) should we make this menu sticky?
b) can we avoid re-rendering the whole page to avoid the scroll-top?
![CleanShot 2022-05-17 at 10 07 15](https://user-images.githubusercontent.com/12430020/168750327-28f45030-2038-4f71-8ea6-b0deb5cd80dd.gif)

cpapazoglou commented on client/my-sites/plugins/categories/index.tsx:
```diff
@@ -17,26 +16,44 @@ export type Category = {
 	separator?: boolean;
 };
 
+const ALLOWED_CATEGORIES = [
```
What's the difference between this `ALLOWED_CATEGORIES` and the one declared in https://github.com/Automattic/wp-calypso/blob/402f24d0d3ec43ddd1d31ea3aaef55945f97bfe3/client/my-sites/plugins/categories/use-categories.tsx#L8 ?

If we need two different lists, can we add a comment right here?

lsl replied:
Can probably ignore, hoping to remove these soon, see first item in: https://github.com/Automattic/wp-calypso/issues/63030

lsl replied:
oh, these are new, should have checked the diff before commenting! Ignore me :smile:

Rolilink replied:
> What's the difference between this ALLOWED_CATEGORIES and the one declared in

@cpapazoglou one is the default one (used by the controller to figure out when to render a category or a plugin) and the other is used to showcase which categories will be render in the menu.

cpapazoglou commented on packages/components/src/responsive-toolbar-group/index.tsx:
```diff
@@ -0,0 +1,215 @@
+								onToggle();
+							} }
+						>
+							More
```
Let's pass this to a translate function.

Rolilink commented:
> * Focus bordering needs hiding

On it! I thought the same about this one.

> * On click animation / behavior need cleaning up - secondary clicks shouldn't react

I also think the same, but doesn't seem to be a blocker could this be a follow-up?

> * Are these button on click animations based on anything else in calypso? Would it make more sense to find something similar elsewhere?

Here we are using the `ToolbarButtonGroup` component from `@wordpress/components` which @SaxonF mentioned here pdh6GB-Qr-p2#comment-1548, I assume we are trying to move to a new design but I can be wrong so I would love Saxon design input on this one. I also think we could launch this and iterate on the design.

> * Dropdown looks quite different to other dropdowns around calypso - can we find something to base this on on for consistency?

I think with Saxon input we can do this in a follow-up.

Rolilink committed:
fixes plugin-browser test

Rolilink committed:
fixes z-index on mobile

Rolilink committed:
removes focus borders

lsl commented:
> b) can we avoid re-rendering the whole page to avoid the scroll-top?
> b) Is a good follow-up ticket.
This is coming from here https://github.com/Automattic/wp-calypso/blob/add/categories-responsive-menu/client/my-sites/plugins/index.js#L88 - can probably just put a hash in the url to avoid `#categories` for now. I'd like to remove that from the index/controller in future but I'm not sure how / when we can do that.

Rolilink committed:
hides responsive menu in mobile

Rolilink committed:
moves ResponsiveToolbarGroup outside of @automattic/components

This is done because ResponsiveToolbarGroup haves a depedency on @wordpress/components which haves a dependency on the window object (https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/resizable-box/resize-tooltip/utils.ts#L12) causing all tests in calypso without jest-environment jsdom to fail.

Rolilink committed:
Fixes E2E test

Rolilink committed:
Merge branch 'trunk' into add/categories-responsive-menu

Rolilink committed:
lint

Rolilink committed:
changing selector on E2E test

Rolilink committed:
adding documentation
